### DISCUSSION TITLE
opeing KDE konsole (and other kde apps) is slow (Hard linking o_tmpfile is slow)

### DISCUSSION THREAD

**Discussion by KoykL** [openzfs/zfs#13441] [Q&A] [answered]
Based on: https://github.com/openzfs/zfs/blob/52bad4f23daaa5f827f802c8d05785a27b80275d/module/os/linux/zfs/zfs_vnops_os.c#L3427
It seems hard linking a o_tmpfile waits for txg completion even when ZIL exists. Therefore, this operation is slow on a non-ssd pool even when ZIL exists. This causes performance problems for many kde apps (e.g. konsole), since they will perform this operation several times on launch (for updating config/session file).

Why is this wait necessary (instead of committing to ZIL)? Is there any way to work around this problem? Thank you for the help!

Example strace:

>openat(AT_FDCWD, "(redacted)/.config", O_RDWR|O_CLOEXEC|O_TMPFILE, 0600) = 11
statx(AT_FDCWD, "(redacted)/.config/konsolerc", AT_STATX_SYNC_AS_STAT|AT_NO_AUTOMOUNT, STATX_ALL, {stx_mask=STATX_ALL|STATX_MNT_ID, stx_attributes=0, stx_mode=S_IFREG|0600, stx_size=1753, ...}) = 0
access("(redacted)/.config/konsolerc", R_OK) = 0
access("(redacted)/.config/konsolerc", W_OK) = 0
access("(redacted)/.config/konsolerc", X_OK) = -1 EACCES (Permission denied)
fchmod(11, 0600)                        = 0
statx(11, "", AT_STATX_SYNC_AS_STAT|AT_NO_AUTOMOUNT|AT_EMPTY_PATH, STATX_ALL, {stx_mask=STATX_ALL|STATX_MNT_ID, stx_attributes=0, stx_mode=S_IFREG|0600, stx_size=0, ...}) = 0
write(11, (redacted), 1760) = 1760
statx(11, "", AT_STATX_SYNC_AS_STAT|AT_NO_AUTOMOUNT|AT_EMPTY_PATH, STATX_ALL, {stx_mask=STATX_ALL|STATX_MNT_ID, stx_attributes=0, stx_mode=S_IFREG|0600, stx_size=1760, ...}) = 0
lseek(11, 0, SEEK_SET)                  = 0
fdatasync(11)                           = 0
linkat(AT_FDCWD, "/proc/self/fd/11", AT_FDCWD, "(redacted)/.config/konsolerc", AT_SYMLINK_FOLLOW) = -1 EEXIST (File exists)
linkat(AT_FDCWD, "/proc/self/fd/11", AT_FDCWD, "(redacted)/.config/konsolerc.OpMaLI", AT_SYMLINK_FOLLOW) = 0 (**pauses here for a while**)
close(11)

**Comment by scineram** [comment, marked as answer]
[Looks like logging tmpfile is not easily done.](https://github.com/openzfs/zfs/pull/4276#issuecomment-248070935)
